Compfiles/CJMO2024P1.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2025 The Compfiles Contributors. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+-/
+
+import Mathlib
+import ProblemExtraction
+
+problem_file { tags := [.Combinatorics] }
+
+/-!
+# Canadian Junior Math Olympiad 2024, Problem 1
+
+https://cms.math.ca/wp-content/uploads/2024/04/cjmo2024-solutions-en.pdf
+
+Centuries ago, the pirate Captain Blackboard buried a vast amount of treasure in a single
+cell of a 2 × 4 grid-structured island. You and your crew have reached the island and have
+brought special treasure detectors to find the cell with the treasure. For each detector, you can
+set it up to scan a specific subgrid [a, b]×[c, d] with 1 ≤ a ≤ b ≤ 2 and 1 ≤ c ≤ d ≤ 4. Running
+the detector will tell you whether the treasure is in the region or not, though it cannot say
+where in the region the treasure was detected. You plan on setting up Q detectors, which may
+only be run simultaneously after all Q detectors are ready. What is the minimum Q required
+to guarantee your crew can determine the location of Blackboard’s legendary treasure?
+-/
+
+open Set
+
+/-- A single cell of the `2 × 4` grid.
+    We index rows with `Fin 2` and columns with `Fin 4`. -/
+abbrev Cell : Type := Fin 2 × Fin 4
+
+/-- An *axis-aligned detector*, defined by the set of cells it scans.
+    (The usual rectangle parameters `[a,b] × [c,d]` are implicit in this set.) -/
+structure Detector where
+  cells : Finset Cell
+
+namespace Detector
+
+/-- Construct a detector from row bounds `a ≤ b` and column bounds `c ≤ d`. -/
+def ofBounds
+    (a b : Fin 2)
+    (c d : Fin 4)
+: Detector where
+  cells := { p : Cell | a ≤ p.1 ∧ p.1 ≤ b ∧ c ≤ p.2 ∧ p.2 ≤ d }
+
+end Detector
+
+/-- The outcome vector produced by *Q* detectors for a hidden treasure cell. -/
+def outcome {Q : ℕ} (D : Fin Q → Detector) (treasure : Cell) : List.Vector Bool Q :=
+  List.Vector.ofFn (fun i ↦ treasure ∈ (D i).cells)
+
+/-- A family of detectors *distinguishes* every treasure location
+    if distinct cells yield distinct outcome vectors. -/
+def Distinguishes {Q : ℕ} (D : Fin Q → Detector) : Prop :=
+  ∀ {c₁ c₂ : Cell}, c₁ ≠ c₂ → outcome D c₁ ≠ outcome D c₂
+
+/-- The minimum number of non-adaptive detectors needed to guarantee
+    the treasure cell can be identified. -/
+noncomputable def minQ : ℕ :=
+  sInf {Q : ℕ | ∃ D : Fin Q → Detector, Distinguishes D}
+
+def D (i : Fin 3): Detector :=
```

Comment:
I think `D` and `D_distinguishes` should be within a `snip` section? See https://github.com/dwrensha/compfiles?tab=readme-ov-file#extracting-problems.